### PR TITLE
Support packaging the project in JDK21

### LINKED ISSUE #11510: [Feature] Response of Temurin JDK 11 EOF at Oct 2024
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

According to Temurin's official page, the JDK 11 should be **End of Availability** **At least Oct 2024**.

Ref to https://adoptium.net/support/
image

To avoid the outdated JDK version for various platforms compiling, we should begin to investigate and prepare the JDK upgrade.
~~Through #11464, I made OAP and UI Webapp running in the JDK21 image. Now, we should decide whether we choose JDK 17(next TLS) or skip it to JDK 21 directly.~~

❗ Updated on Nov. 8th, JDK 21 and/or JRE 21 as runtime are not available for now.

> Groovy 3(we are using) doesn't support JRE 21, as well as the latest stable version Groovy 4.
JRE 21 support is in the Grooxy 5, but it is at alpha stage only.
So, the current status is, that we could change the base image to Zulu is Temurin is not updated next year.

FYI all @apache/skywalking-committers If you have any preference, please comment here and explain the reasons, thanks.

___
### Oracle OpenJDK
According to Zipkin server(v3 is sharing SkyWalking kernel), JDK 17 is required. 
And Oracle OpenJDK 11/17/21 has many years before EOF. Although we don't use Oracle OpenJDK, but this is a good reference before taking actions.

image

### Azul Zulu OpenJDK
Zulu OpenJDK has a longer JDK11 support too.
image

### Are you willing to submit a pull request to implement this on your own?

- [X] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by wu-sheng**
- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #11510.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

Support packaging the project in JDK21. The compiler source and target remain in JDK11. Runtime JRE requirements are JRE 11 and 17. JRE 21 is not supported due to Groovy core limitation.

cc @apache/skywalking-committers I think we are fine with new JDK 17 and 21 packaging, and the codes are still running in JRE 11 and 17.

cc @tisonkun The conclusion is above.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -284,6 +285,28 @@
...
+                <groupId>net.bytebuddy</groupId>
+                <artifactId>byte-buddy</artifactId>
+                <version>${byte-buddy.version}</version>
+                <scope>test</scope>
+            </dependency>
+            <dependency>
+                <groupId>net.bytebuddy</groupId>
+                <artifactId>byte-buddy-agent</artifactId>
+                <version>${byte-buddy.version}</version>
+                <scope>test</scope>
```
Comment (wu-sheng) [inline pom.xml:309]: `byte-buddy.version`=1.14.9 is the key to support JRE21, as mockito and system-stubs-jupiter are using the agent mechanism for the test mock.

**Comment by kezhenxu94** [comment]
I'm not sure what's the point to compile/package the code with Java 21 if it cannot actually run on Java 21.

**Comment by wu-sheng** [comment]
> compile/package the code with Java 21 if it cannot actually run on Java 21.

There is not a special meaning. I just get the codes working, and prepare for longer term that once Groovy 5 could work on  JRE 21, we should be able to allow JDK 21 as local dev/test env.

File: docs/en/guides/How-to-build.md [hunk 2]
```diff
@@ -16,7 +16,7 @@ If you need to execute build behind the proxy, edit the *.mvn/jvm.config* and se
 ```
 
 ### Building from GitHub
-1. Prepare git, JDK11+, and Maven 3.6+.
+1. Prepare git, JDK 11/17/21(LTS versions), and Maven 3.6+.
```
Comment (kezhenxu94) [inline docs/en/guides/How-to-build.md:19]: Adding 21 here makes no sense to me. Packages built with Java 21 is totally unusable, it's a waste of time for the users, and it adds up unnecessary potential issues that report this problem.

**Review by kezhenxu94 — COMMENTED** [review]
> > compile/package the code with Java 21 if it cannot actually run on Java 21.
> 
> 
> 
> There is not a special meaning. I just get the codes working, and prepare for longer term that once Groovy 5 could work on  JRE 21, we should be able to allow JDK 21 as local dev/test env.

If so I would suggest either:

- not to add Java 21 anywhere in the doc that is user oriented, it's rather confusing to say we can compile the project under Java 21 but it cannot actually run under Java 21;
- or, configure maven to produce lower version byte-codes so it compiles with 21 but can be run under 11 or 17, this is a common practice for Java libraries that are developed in a higher JDK but can be run in lower JREs, although our project is not a library. 


Either way, I agree we can add 21 to the test matrix and fix existing issues as you have already done 👍

**Comment by wu-sheng** [comment]
> or, configure maven to produce lower version byte-codes so it compiles with 21 but can be run under 11 or 17, this is a common practice for Java libraries that are developed in a higher JDK but can be run in lower JREs, although our project is not a library.

I don't change compiler target. So, yes, the codes are still in the JDK11 level byte codes.

File: docs/en/guides/How-to-build.md (hunk 2 above)
Comment (wu-sheng) [reply docs/en/guides/How-to-build.md:19]: 21 has been removed.

**Review by kezhenxu94 — APPROVED** [review]

**Pull request merged**